### PR TITLE
Sync ``libasr`` from LFortran

### PR DESCRIPTION
**Pull request by czgdp1807**

### CODE REVIEW

File: src/libasr/utils.h [hunk 1]
```diff
@@ -32,41 +41,37 @@ struct CompilerOptions {
...
     std::string error_format = "human";
     bool new_parser = false;
     bool implicit_typing = false;
     bool implicit_interface = false;
+    bool rtlib = false;
     std::string target = "";
-    std::string arg_o = "";
-    bool emit_debug_info = false;
-    bool emit_debug_line_column = false;
-    std::string import_path = "";
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/utils.h:44]: We need to revert this.

File: src/libasr/runtime/lfortran_intrinsics.h [hunk 2]
```diff
@@ -40,7 +40,6 @@ LFORTRAN_API double _lfortran_random();
 LFORTRAN_API int _lfortran_randrange(int lower, int upper);
 LFORTRAN_API int _lfortran_random_int(int lower, int upper);
 LFORTRAN_API void _lfortran_printf(const char* format, ...);
-LFORTRAN_API void _lcompilers_print_error(const char* format, ...);
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/runtime/lfortran_intrinsics.h:43]: ..

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 3]
```diff
@@ -56,15 +56,6 @@ LFORTRAN_API void _lfortran_printf(const char* format, ...)
...
 }
 
-LFORTRAN_API void _lcompilers_print_error(const char* format, ...)
-{
-    va_list args;
-    va_start(args, format);
-    vfprintf(stderr, format, args);
-    fflush(stderr);
-    va_end(args);
-}
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/runtime/lfortran_intrinsics.c:66]: ..

**Comment by czgdp1807** [comment]
@Thirumalai-Shaktivel I am going through the diff right now. When I will be done I will ping you and others for a second round of inspection. Thanks.

**Comment by Thirumalai-Shaktivel** [comment]
Cool, it is highly difficult to sync the changes. Thanks for doing this!

**Comment by czgdp1807** [comment]
@certik @Shaikh-Ubaid @lucifer1004 @konradha @Pranavchiku @Smit-create @Thirumalai-Shaktivel Please go through the diff and let me know if there's any change that got reverted. I will restore it.

**Comment by czgdp1807** [comment]
Windows test fail, probably due to the changes in `src/libasr/codegen/KaleidoscopeJIT.h`. @certik Can you check what needs to be done there?

**Review by Shaikh-Ubaid — APPROVED** [review]
The changes look fine to me. Thank you so much for this.

File: src/libasr/pass/pass_list_concat.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
+#include <libasr/asr.h>
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/pass_list_concat.cpp:1]: This file was renamed to pass_list_expr.cpp, so this can be removed

**Review by Thirumalai-Shaktivel — APPROVED** [review]
Apart from the above change, this looks good to me as well.

**Review by Pranavchiku — APPROVED** [review]
Looks good to me!

**Review by Smit-create — APPROVED** [review]
Thanks for this!

**Comment by certik** [comment]
I am reviewing it.

**Comment by czgdp1807** [comment]
Windows test fail. I think its something related to `src/libasr/codegen/KaleidoscopeJIT.h`. That needs to be worked out. Rest everything is done from my end. @certik I will give you the push access to my fork so that you can make any changes to make Windows tests work.

Also note that WASM shouldn't define functions by topologically sorting them simply because LFortran has some tests where functions depend cyclically on each other and hence they fail. So at least to move things forwarded I reverted to first declaring the functions in WASM backend and then defining them. We need to figure out a general solution for cyclic dependencies later.

**Review by certik — APPROVED** [review]
I went over everything, I haven't noticed any issues. Thanks!

**Comment by certik** [comment]
The Windows failures should be fixed in https://github.com/lcompilers/lpython/pull/1314, they are caused by the new LLVM JIT being more strict about what you can do, so some tests that worked with the old JIT do not work anymore.

**Comment by certik** [comment]
Closing in favor of https://github.com/lcompilers/lpython/pull/1314.

**Pull request closed without merging**